### PR TITLE
Allow field access on structural constructors

### LINKED ISSUE #20238: Cannot use field/index/invocation access on grouped expressions
**Description:**
#20234 Partially fixes the issue. However, still cannot use expressions within the group (i.e ("A" + x).length(), (<PersonObj>(dataMap)["a"]).getName()) due to current grammar limitations.

**Steps to reproduce:**
```ballerina
# Fails because `("string " + x)` is not supported within group invocation.
 function testGroupedLangLibInvocationRef() returns boolean {
     string x = "value";
     return (("string " + x)).length() == 12;
 }

# Fails because `<PersonObj>(dataMap)["a"]` is not supported within group invocation.
 function testNestedGroupedInvocationRef() returns boolean {
     PersonObj p = new;
     map<PersonObj> dataMap = {"a": p};
     return p.getName() === (<PersonObj>(dataMap)["a"]).getName();
 }

# Fails because `isNaN()` invocation on (0.0/0.0) expression is not supported.
function testGroupedBuiltInInvocationRef() returns boolean {
    float f = 0.0/0.0;
    return f.isNaN() && (f).isNaN() && (0.0/0.0).isNaN();
}
```

**Affected Versions:**
1.0.5, 1.1.0

**Related Issues:**
- https://github.com/ballerina-platform/ballerina-lang/issues/18298
- https://github.com/ballerina-platform/ballerina-lang/issues/19020

### LINKED ISSUE #25219: ClassCast error at compile time for a query expression usage
**Description:**

**Steps to reproduce:**
```ballerina
function foo() {
    EmployeeEntity entity = {id: 1232, fname: "S", lname: "J", age: 30, addr: {line1: "N", line2: "561", city: "S", state: "C", postalCode: "95"}};
    EmployeeView emp = from var {fname, lname, age, addr:{city, state, postalCode}} in [entity]
                        select {fname, lname, age, addr:{city, state, postalCode}}[0];
    io:println(emp);
}

type EmployeeEntity record {
    int id;
    string fname;
    string lname;
    int age;
    AddressEntity addr;
};

type AddressEntity record {
    string line1;
    string line2;
    string city;
    string state;
    string postalCode;
};

type EmployeeView record {|
    string fname;
    string lname;
    int age;
    AddressView addr;
|};

type AddressView record {|
    string city;
    string state;
    string postalCode;
|};
```
[ballerina-internal.log](https://github.com/ballerina-platform/ballerina-lang/files/5059636/ballerina-internal.log)

**Affected Versions:**
Swan Lake Preview 2

### LINKED ISSUE #26206: BAD-SAD when field access expr is used on a non variable reference
**Description:**
```ballerina
type R record {
    int i;
};

public function main() {
    R r = {i : 0};
    int i = {b}.i;
}
```

```ballerina
type R record {
    int i;
};

public function main() {
    R r = {i : 0};
    int i = (r+b).i;
}
```

ballerina-internal.log for both programs
```
[2020-10-05 19:58:49,064] SEVERE {b7a.log.crash} - org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral cannot be cast to org.wso2.ballerinalang.compiler.tree.expressions.BLangVariableReference 
java.lang.ClassCastException: org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral cannot be cast to org.wso2.ballerinalang.compiler.tree.expressions.BLangVariableReference
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2461)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangFieldBasedAccess.accept(BLangFieldBasedAccess.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:361)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:340)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:663)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2751)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1678)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:2755)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:374)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:346)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2751)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:269)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:158)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:243)
	at org.wso2.ballerinalang.compiler.CompilerDriver.typeCheck(CompilerDriver.java:329)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:258)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:250)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:142)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:176)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:220)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:101)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:109)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:58)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:297)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)
 
[2020-10-05 19:59:16,525] SEVERE {b7a.log.crash} - org.wso2.ballerinalang.compiler.tree.expressions.BLangBinaryExpr cannot be cast to org.wso2.ballerinalang.compiler.tree.expressions.BLangVariableReference 
java.lang.ClassCastException: org.wso2.ballerinalang.compiler.tree.expressions.BLangBinaryExpr cannot be cast to org.wso2.ballerinalang.compiler.tree.expressions.BLangVariableReference
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2461)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangFieldBasedAccess.accept(BLangFieldBasedAccess.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:361)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:340)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:663)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2751)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1678)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:2755)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:374)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:346)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2791)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:2759)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:2751)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:269)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:158)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:243)
	at org.wso2.ballerinalang.compiler.CompilerDriver.typeCheck(CompilerDriver.java:329)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:258)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:250)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:142)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:176)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:220)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:101)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:109)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:58)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:297)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)

```

**Steps to reproduce:**

**Affected Versions:**
SLP-4
**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
Purpose of this PR is to $title.

Fixes #26206
Fixes #20238
Fixes #25219

Note: This PR will change the inheritance hierarchy of BLangAccessExpression and BLangVariableReference. According to (https://ballerina.io/spec/lang/master/#section_7.14.1) var-ref, field access, and member access should have a single abstraction.

```
lvexpr :=
   variable-reference-lvexpr
   | field-access-lvexpr
   | member-access-lvexpr
```

## Approach
n/a

## Samples
n/a

## Related Issues/PRs
- #18298
- #19020
- #20234
- #20238
- #24603
- #25219
- #26206

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
  * @see BLangJSONArrayLiteral
  * @since 1.0
  */
-public class BLangListConstructorExpr extends BLangExpression implements ListConstructorExprNode {
+public class BLangListConstructorExpr extends BLangAccessExpression implements ListConstructorExprNode {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java:38]: I'm not sure if we should be marking these as `BLangAccessExpression`s. These aren't access expressions, right? They are the expressions on which the access is done?

Shouldn't we fix the visits for field/member access instead? To consider any `BLangExpression` rather than expecting a `BLangVariableReference` (https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L2465)?

Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java:38]: I have a mixed opinion on this. But, if we have any better way as @MaryamZi mentioned, then let's explore that and see.

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java:38]: Is it ok to introduce another abstraction with `lhsVar`, `compoundAssignmentLhsVar`, etc..., and expect that? Or is it possible to get rid of analyzing them all together, given that we need to allow all the BLangExpression? Or do we have to analyze case by case (i.e BLangVariableReferences, Structural Constructs, etc...)?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java:38]: I believe we still need to preserve information regarding whether or not certain expressions are on the LHS to support/restrict certain scenarios.

```ballerina
type Foo record {|
    int i?;
|};

public function main() {
    Foo f = {};

    f.i = 1; // field access for an optional field on the LHS - valid
    int? i = f.i; // field access for an optional field on the RHS - invalid

    int? j = (<Foo> {})?.i; // valid

    (<Foo> {}).i = 2; // invalid, LHS is not a variable-reference-lvexpr, field-access-lvexpr, nor a member-access-lvexpr
}
```

The spec has this concept of lvalues (https://ballerina.io/spec/lang/master/#section_7.14.1) and an `lvexpr` is defined as 

```
lvexpr :=
   variable-reference-lvexpr
   | field-access-lvexpr
   | member-access-lvexpr
```

Thinking out loud, `lhsVar`, `compoundAssignmentLhsVar`, etc. may thus only be relevant for `variable-reference`, `field-access`, and `member-access`. 

I think we should look into the possibility of basing the abstraction on this.

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangListConstructorExpr.java:38]: > Thinking out loud, lhsVar, compoundAssignmentLhsVar, etc. may thus only be relevant for variable-reference, field-access, and member-access.

@MaryamZi, I agree, but with the current implementation, `BLangAccessExpression` extends `BLangVariableReference` and both `BLangFieldBasedAccess`, `BLangIndexBasedAccess` extends `BLangAccessExpression`. Since, information like `lhsVar` & `compoundAssignmentLhsVar` belongs to `BLangVariableReference`, there's already a such abstraction. However, it's not only limited to `lvexpr`, but inherits by each and every expression which descendant from `BLangVariableReference`. Is that acceptable?

**Review by mohanvive — DISMISSED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -2367,7 +2367,7 @@ private void createVarRefAssignmentStmts(BLangTupleVarRef parentTupleVariable, B
      * This method creates a assignment statement and assigns and array expression based on the given indexExpr.
      *
      */
-    private void createSimpleVarRefAssignmentStmt(BLangVariableReference simpleVarRef, BLangBlockStmt parentBlockStmt,
+    private void createSimpleVarRefAssignmentStmt(BLangExpression simpleVarRef, BLangBlockStmt parentBlockStmt,
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:2370]: Shall we rename this method and first param, as this function now support more than simpleVariableReferences

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed in 278547735197ee003300179d272056e1b9251534

**Review by mohanvive — DISMISSED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ASTBuilderUtil.java [hunk 3]
```diff
@@ -724,19 +723,19 @@ public static BLangMatchExpression createMatchExpression(BLangExpression expr) {
         return matchExpr;
     }
 
-    public static BLangFieldBasedAccess createFieldAccessExpr(BLangVariableReference varRef, BLangIdentifier field) {
+    public static BLangFieldBasedAccess createFieldAccessExpr(BLangExpression varRef, BLangIdentifier field) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ASTBuilderUtil.java:726]: Can we accept `BLangAccessibleExpression`s here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -2367,25 +2367,25 @@ private void createVarRefAssignmentStmts(BLangTupleVarRef parentTupleVariable, B
      * This method creates a assignment statement and assigns and array expression based on the given indexExpr.
      *
      */
-    private void createSimpleVarRefAssignmentStmt(BLangVariableReference simpleVarRef, BLangBlockStmt parentBlockStmt,
-                                                  BLangExpression indexExpr, BVarSymbol tupleVarSymbol,
-                                                  BLangIndexBasedAccess parentArrayAccessExpr) {
+    private void createAssignmentStmt(BLangAccessExpression accessExpression, BLangBlockStmt parentBlockStmt,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:2370]: Shouldn't this be a `BLangAccessibleExpression`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 5]
```diff
@@ -2462,11 +2463,14 @@ private boolean isFunctionArgument(BSymbol symbol, List<BLangSimpleVariable> par
...
 
     public void visit(BLangFieldBasedAccess fieldAccessExpr) {
-        // First analyze the variable reference expression.
-        ((BLangVariableReference) fieldAccessExpr.expr).lhsVar = fieldAccessExpr.lhsVar;
-        ((BLangVariableReference) fieldAccessExpr.expr).compoundAssignmentLhsVar =
-                fieldAccessExpr.compoundAssignmentLhsVar;
-        BType varRefType = getTypeOfExprInFieldAccess(fieldAccessExpr.expr);
+        // First analyze the accessible expression.
+        BLangExpression containerExpression = fieldAccessExpr.expr;
+        if (containerExpression instanceof BLangAccessibleExpression) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2468]: Shouldn't this always be true? If not, shouldn't we log an error or return?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessExpression.java [hunk 6]
```diff
@@ -26,12 +25,13 @@
...
-public abstract class BLangAccessExpression extends BLangVariableReference implements VariableReferenceNode {
+public abstract class BLangAccessExpression extends BLangAccessibleExpression {
 
     public BLangExpression expr;
     public boolean optionalFieldAccess = false;
     public boolean errorSafeNavigation = false;
     public boolean nilSafeNavigation = false;
     public BType originalType;
     public boolean leafNode;
+
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessExpression.java:36]: Unnecessary change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessibleExpression.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessibleExpression.java:2]:
```suggestion
 *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessibleExpression.java]: Fixed in 004d3a3

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessExpression.java (hunk 6 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangAccessExpression.java]: Fixed in 004d3a3

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 5 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Fixed in 004d3a3

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 4 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed in 004d3a3

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ASTBuilderUtil.java (hunk 3 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ASTBuilderUtil.java]: Fixed in 004d3a3

**Review by gimantha — APPROVED** [review]
